### PR TITLE
Export object_store integration tests

### LINKED ISSUE #5685: feat object_store: moving tests from src/ to a tests/ folder and enabling access to test functions for enabling a shared integration test suite
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Hi young dev here! 

I am currently busy doing an external implementation (or to phrase it better, I am essential refactoring the [datafusion-contrib/datafusion-objectstore-hdfs](https://github.com/datafusion-contrib/datafusion-objectstore-hdfs) repository to support the latest 0.10 object store interface. 

While my src implementation is complete, I am still lacking the tests. While looking at the implementation of the `object_store` interface in this repository, I've noticed that most of tests in ObjectStore implementations (`Local`, `Azure` etc.) call test functions defined in `object_store/src/lib.rs`. These tests offer a crate visibility, so even thou my implementation would benefit from it, I can't access it.

After a brief discussion with @tustvold, he suggested to generalise my need to create a suite of shared integration tests across object_store implementations.

**Describe the solution you'd like**

This solution comprises of mainly two points
- [ ] Move the tests from the source files to a mirror into `/tests` folder
- [ ] Change the visibility (from `pub(crate)` to `pub`) of test functions, ensuring this change is not in conflict with private accesses.

**Describe alternatives you've considered**

Alternatives to this feature would be to decouple the tests (completely refactor what is the current status of the repository) leading to code repetition (tests are doing a similar task, even for different implementations.

If there is no will of maintaining these tests (making them public), any external implementation of the object_store interface should resort to a copy-paste of the functions currently present (code duplication).

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #5685.

# Rationale for this change
 

The object_store crate has quite a comprehensive test suite, to better enable third-party implementations we can export this for external consumption.

I debated making this a separate crate, but I felt a feature flag was more discoverable, and avoids all the complexities associated with needing to create a separate object_store workspace to contain both crates, and then release both in sync.

# What changes are included in this PR?

# Are there any user-facing changes?

No

### CODE REVIEW

File: object_store/src/lib.rs [hunk 1]
```diff
@@ -1299,1054 +1304,19 @@ mod test_util {
...
+        store: &'a dyn ObjectStore,
+        path_str: &str,
+    ) -> BoxStream<'a, Result<ObjectMeta>> {
+        let path = Path::from(path_str);
+        store.list(Some(&path))
     }
 
     #[cfg(any(feature = "azure", feature = "aws"))]
-    pub(crate) async fn signing<T>(integration: &T)
+    pub async fn signing<T>(integration: &T)
```
Comment (tustvold) [inline object_store/src/lib.rs:1317]: I opted to leave signing and tagging behind as they're relatively niche, and created some funky feature flag/dependency shenanigans.

File: object_store/src/parse.rs [hunk 2]
```diff
@@ -25,15 +25,9 @@ use url::Url;
 
 #[derive(Debug, Snafu)]
 enum Error {
-    #[snafu(display("Unable to convert URL \"{}\" to filesystem path", url))]
```
Comment (tustvold) [inline object_store/src/parse.rs:28]: Unrelated clippy lint fix

**Comment by tustvold** [comment]
Looks like a new clippy is better at finding unused code, so rolled that into this PR

File: object_store/src/parse.rs (hunk 2 above)
Comment (alamb) [reply object_store/src/parse.rs:28]: is this an API change (as the type enums have changed?)

**Review by alamb — COMMENTED** [review]
Thank you @tustvold 

I think we should:
1. Document the new feature flag (along with an example of how to run the integration tests -- I didn't see this anywhere but may have missed it) Perhaps add to docs [here](https://docs.rs/object_store/latest/object_store/#available-objectstore-implementations)
2. Figure out if this is an API change or not, and if it is an API change split those out from the extraction to a new feature flag (to allow finer grained control of when breaking changes are needed)

This is a really neat PR BTW

File: object_store/src/parse.rs (hunk 2 above)
Comment (tustvold) [reply object_store/src/parse.rs:28]: These enumerations are module private

Comment (alamb) [reply object_store/src/parse.rs:28]: I double checked that this is not visible https://docs.rs/object_store/latest/object_store/path/enum.Error.html?search=InvalidUrl

File: object_store/src/aws/builder.rs [hunk 3]
```diff
@@ -70,21 +70,9 @@ enum Error {
     #[snafu(display("Configuration key: '{}' is not known.", key))]
     UnknownConfigurationKey { key: String },
 
-    #[snafu(display("Bucket '{}' not found", bucket))]
-    BucketNotFound { bucket: String },
```
Comment (alamb) [inline object_store/src/aws/builder.rs:74]: double checked it is not exposed: https://docs.rs/object_store/latest/object_store/path/enum.Error.html?search=BucketNotFound

**Review by alamb — APPROVED** [review]
I still think documenting how to use this integration test is really important, but I think it could be done as a follow on PR too.

Thanks @tustvold

**Pull request merged**